[REEF-1388] Fix RunningTask to be sent for short-lived .NET tasks

tcNickolas opened the pull request:
This change:
 * sends RUNNING message immediately before Task.call(), instead of
   relying on time-based heartbeat to report task in Running state.
 * removes Thread.Sleep() calls from tests which don't need them any more.
 * fixes tests which rely on running status improperly.

JIRA:
  [REEF-1388](https://issues.apache.org/jira/browse/REEF-1388)

Pull request:
  This closes #

tcNickolas commented:
I got a clean AppVeyor run of almost this code at https://ci.appveyor.com/project/tcNickolas/reef/build/118-REEF-1388 (it differs in some debug changes in bridge test which I've reverted since). On other runs I'm getting transient failures of TestEvaluatorWithActiveContextImmediatePoison (REEF-1406) and TestEnforceCloseTaskOnLocalRuntime (unexplained). Let's 

@markusweimer Could you please test this PR on your laptop which allows to repro AppVeyor failures? I can't repro failures on my machine.

afchung commented on lang/cs/Org.Apache.REEF.Evaluator.Tests/ContextRuntimeTests.cs:
```diff
@@ -184,7 +186,17 @@ public void TestUnableToSpawnChildWhileTaskIsRunning()
                     Assert.True(contextRuntime.GetTaskStatus().IsPresent());
-                    Assert.Equal(contextRuntime.GetTaskStatus().Value.state, State.RUNNING);
+
+                    // wait for the task to start
```
Could you use bind volatile to bind a `CountDownEvent` instead of using a timeout?

afchung replied:
Same for all other test cases.

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestFailedEvaluatorEventHandler.cs:
```diff
@@ -143,15 +130,13 @@ public void Dispose()
+                Logger.Log(Level.Info, "Entered FailEvaluatorTask");
+
+                // need to sleep to allow Java code to get heartbeat with information about task
+                Thread.Sleep(100);
```
If you move `Heartbeat()` prior to `_taskLifeCycle.Start()` in `SetInit()` in `TaskStatus.cs`, would that solve the problem?

tcNickolas replied:
No, still get the same test failure (without sleep).

afchung replied:
Hmm, so what are the symptoms? The Driver doesn't see that the Task exists before the Evaluator fails?

afchung replied:
I would assume that sending a heartbeat in `SetInit()` should inform the Driver that the Task exists. Given that a Task should be able to fail before turning `Running`, this may be a bug in of itself.

tcNickolas replied:
Yes, the time for task information to propagate to Java code is more than time for .NET code to fail.

I've just tried old version of the test, with `CountdownEvent` and `RunningTask`, and it passes both locally and [on AppVeyor](https://ci.appveyor.com/project/tcNickolas/reef/build/136-REEF-1388-t/tests). Most likely I've done this change when trying to fix tests before deadlock in reporting was resolved. 
I'll revert this test (which also addresses the other comments on it)

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs:
```diff
@@ -120,16 +122,23 @@ public void OnNext(ICompletedTask value)
-                // Suspend the first instance of the Task.
-                value.Suspend(Encoding.UTF8.GetBytes(SuspendMessageFromDriver));
+                ++_taskIndex;
+                if (_taskIndex == 1)
```
Why not use a `bool` here?

afchung commented on lang/cs/Org.Apache.REEF.Examples/Tasks/HelloTask/HelloTask.cs:
```diff
@@ -66,7 +66,7 @@ public byte[] Call(byte[] memento)
             }
             PrintGuestList();
-            Thread.Sleep(5 * 1000);
+            Thread.Sleep(1000);
```
What is this for?

tcNickolas replied:
It was needed at one of previous iterations, but now tests pass both locally and [on AppVeyor](https://ci.appveyor.com/project/tcNickolas/reef/build/134-REEF-1388-t) without this, so probably not needed any more

afchung commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs:
```diff
@@ -440,15 +440,7 @@ public Optional<TaskStatusProto> GetTaskStatus()
-                    return Optional<TaskStatusProto>.Of(taskStatusProto);
-                }
-
-                var e = new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, "Task state must be RUNNING, but instead is in {0} state", taskStatusProto.state));
```
@jwang98052 would you know why this was here before? Is it okay to remove this?

jwang98052 replied:
I am not sure. 

From the comments of the caller, it says "the TaskStatusProto of the currently running task, if there is any", "the state of the running Task, if one is running." Meaning it expects the task is running. 

Do we ever see this exception ever happen? I didn't.

tcNickolas replied:
We have observed in this part of the code comments which do not correspond to what the code actually does (for example, the `Send heartbeat such that user receives a TaskRunning message.` which started this whole fix). I wouldn't read too much into the comments here, because they can describe incorrect behavior of code. I don't think a method called `GetTaskStatus` should throw exception if the status is anything but running.

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestFailedEvaluatorEventHandler.cs:
```diff
@@ -91,15 +86,9 @@ public void OnNext(IAllocatedEvaluator value)
                     .Build());
             }
 
-            public void OnNext(IRunningTask value)
```
Why removing running task handler? Does it still not work?

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestFailedEvaluatorEventHandler.cs:
```diff
@@ -143,15 +130,13 @@ public void Dispose()
                 return null;
             }
-
-            public void Handle(IDriverMessage message)
```
Not have enough time to receive driver message?

jwang98052 commented:
I completed a pass. 

I also applied the change in this PR to my IMRU fault tolerant test in my private branch, the test passed. Before the change, the test failed to receive running task event.

tcNickolas commented:
I've addressed comments and updated the PR. Successful AppVeyor run is https://ci.appveyor.com/project/tcNickolas/reef/build/136-REEF-1388-t (Apache AppVeyor seems to be stuck systematically by this time of day)

afchung commented on lang/cs/Org.Apache.REEF.Examples/Tasks/HelloTask/HelloTask.cs:
```diff
@@ -66,7 +66,7 @@ public byte[] Call(byte[] memento)
             }
             PrintGuestList();
-            Thread.Sleep(5 * 1000);
+            ////Thread.Sleep(1000);
```
Please remove commented out code.

afchung commented:
Looks good to me.

tcNickolas commented:
I've updated PR to remove commented out code

afchung commented:
+1 for merge. Will wait for @jwang98052 to sign off.

jwang98052 commented:
LGTM.

afchung commented:
I'll test and merge.
